bumble/hid.py

```diff
@@ -0,0 +1,301 @@
+# limitations under the License.
+
+# -----------------------------------------------------------------------------
+# Imports
+# -----------------------------------------------------------------------------
+from __future__ import annotations
+from dataclasses import dataclass
+import logging
+import asyncio
+import enum
+
+from pyee import EventEmitter
+from typing import Optional, Tuple, Callable, Dict, Union, TYPE_CHECKING
+
+from . import core, l2cap  # type: ignore
+from .colors import color  # type: ignore
+from .core import BT_BR_EDR_TRANSPORT, InvalidStateError, ProtocolError  # type: ignore
+
+if TYPE_CHECKING:
+    from bumble.device import Device, Connection
+# -----------------------------------------------------------------------------
+# Logging
+# -----------------------------------------------------------------------------
+logger = logging.getLogger(__name__)
+
+
+# -----------------------------------------------------------------------------
+# Constants
+# -----------------------------------------------------------------------------
+# fmt: off
+HID_CONTROL_PSM            = 0x0011
+HID_INTERRUPT_PSM          = 0x0013
+
+
+class Message():
+
+    # Report types
+    class ReportType(enum.IntEnum):
+        OTHER_REPORT        = 0x00
+        INPUT_REPORT        = 0x01
+        OUTPUT_REPORT       = 0x02
+        FEATURE_REPORT      = 0x03
+
+    # Handshake parameters
+    class Handshake(enum.IntEnum):
+        SUCCESSFUL              = 0x00
+        NOT_READY               = 0x01
+        ERR_INVALID_REPORT_ID   = 0x02
+        ERR_UNSUPPORTED_REQUEST = 0x03
+        ERR_UNKNOWN             = 0x0E
+        ERR_FATAL               = 0x0F
+
+    #Message Type
+    class MessageType(enum.IntEnum):
+        HANDSHAKE           = 0x00
+        CONTROL             = 0x01
+        GET_REPORT          = 0x04
+        SET_REPORT          = 0x05
+        GET_PROTOCOL        = 0x06
+        SET_PROTOCOL        = 0x07
+        DATA                = 0x0A
+
+    # Protocol modes
+    class ProtocolMode(enum.IntEnum):
+        BOOT_PROTOCOL      = 0x00
+        REPORT_PROTOCOL    = 0x01
+
+    # Control Operations
+    class ControlCommand(enum.IntEnum):
+        SUSPEND              = 0x03
+        EXIT_SUSPEND         = 0x04
+        VIRTUAL_CABLE_UNPLUG = 0x05
+
+
+# HIDP messages
+@dataclass
+class GetReportMessage(Message):
+    report_type : int
+    report_id : int
+    buffer_size : int
+
+    def __bytes__(self) -> bytes:
+        if(self.report_type == Message.ReportType.OTHER_REPORT):
+            param = self.report_type
+        else:
+            param = 0x08 | self.report_type
+        header = ((Message.MessageType.GET_REPORT << 4) | param)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        packet_bytes.append(self.report_id)
+        packet_bytes.extend([(self.buffer_size & 0xff), ((self.buffer_size >> 8) & 0xff)])
+        return bytes(packet_bytes)
+@dataclass
```

Comment:
there should be 2 blank lines between top-level function and class definitions (see [pep8)](https://peps.python.org/pep-0008/#blank-lines)

Reply:
Done